File: Mcma.Core/ServiceClient.cs
using System.Collections.Generic;
using System.Linq;

namespace Mcma.Core
{
    public class ServiceClient
    {
        public ServiceClient(Service service, IMcmaAuthenticatorProvider authProvider = null)
        {
            Data = service;

            ResourcesByType =
                service.Resources != null
                    ? service.Resources.ToDictionary(r => r.ResourceType, r => new ResourceEndpointClient(r, authProvider, service.AuthType, service.AuthContext))
                    : new Dictionary<string, ResourceEndpointClient>();
        }

        public Service Data { get; }

        private IDictionary<string, ResourceEndpointClient> ResourcesByType { get; }

        public IEnumerable<ResourceEndpointClient> Resources => ResourcesByType.Values;

        public bool HasResourceEndpoint(string resourceType)
            => ResourcesByType.ContainsKey(resourceType);

        public bool HasResourceEndpoint<T>()
            => HasResourceEndpoint(typeof(T).Name);

        public ResourceEndpointClient GetResourceEndpoint(string resourceType)
            => ResourcesByType.ContainsKey(resourceType) ? ResourcesByType[resourceType] : null;
            

        public ResourceEndpointClient GetResourceEndpoint<T>()
            => GetResourceEndpoint(typeof(T).Name);
    }
}

File: nodejs/base/client/lib/resources/service-client.ts
import { ServiceProperties } from "@mcma/core";
import { ResourceEndpointClient } from "./resource-endpoint-client";
import { AuthProvider } from "../auth/auth-provider";

export class ServiceClient {
    private endpointsMap: { [key: string]: ResourceEndpointClient } = {};

    constructor(service: ServiceProperties, authProvider: AuthProvider) {

        for (const resourceEndpoint of service.resources) {
            this.endpointsMap[resourceEndpoint.resourceType] =
                new ResourceEndpointClient(resourceEndpoint, authProvider, service.authType, service.authContext);
        }
    }

    private getResourceTypeName(resourceType: string | Function): string {
        return typeof resourceType === "function" && resourceType.name ? resourceType.name : resourceType as string;
    }

    hasResourceEndpoint(resourceType: string | Function): boolean {
        return this.endpointsMap[this.getResourceTypeName(resourceType)] !== undefined;
    }

    getResourceEndpointClient(resourceType: string | Function): ResourceEndpointClient | undefined {
        return this.endpointsMap[this.getResourceTypeName(resourceType)];
    }

    getAllResourceEndpointClients(): ResourceEndpointClient[] {
        return Object.keys(this.endpointsMap).map(k => this.endpointsMap[k]);
    }
}